src/trace/trace.cpp

```diff
@@ -332,42 +347,53 @@ void Trace::add_lo2s_property(const std::string& name, const std::string& value)
 {
     auto name = (boost::format("thread %d") % tid).str();
-    auto location =
-        locations_.emplace(location_ref(), intern(name), location_groups_process_.at(pid),
-                           otf2::definition::location::location_type::cpu_thread);
-    process_comm_groups_.at(pid).add_member(location);
-    return archive()(location);
-}
 
+    // As the tid is unique in this context, create only one writer/location per tid
+    auto location = thread_locations_.emplace(
+        std::piecewise_construct, std::forward_as_tuple(tid),
+        std::forward_as_tuple(location_ref(), intern(name), location_groups_process_.at(pid),
+                              otf2::definition::location::location_type::cpu_thread));
+    process_comm_groups_.at(pid).add_member(location.first->second);
+
+    return archive()(location.first->second);
+}
 otf2::writer::local& Trace::cpu_writer(int cpuid)
 {
-    auto name = intern((boost::format("cpu %d") % cpuid).str());
+    auto name = (boost::format("cpu %d") % cpuid).str();
+
+    // As CPU IDs are unique in this context, create only one writer/location per
+    // CPU ID
     auto r_group = location_groups_cpu_.emplace(
         std::piecewise_construct, std::forward_as_tuple(cpuid),
-        std::forward_as_tuple(location_group_ref(), name,
+        std::forward_as_tuple(location_group_ref(), intern(name),
                               otf2::definition::location_group::location_group_type::unknown,
                               system_tree_cpu_nodes_.at(cpuid)));
     const auto& group = r_group.first->second;
 
-    auto location = locations_.emplace(location_ref(), name, group,
-                                       otf2::definition::location::location_type::cpu_thread);
-    return archive()(location);
+    auto location = cpu_locations_.emplace(
+        std::piecewise_construct, std::forward_as_tuple(cpuid),
+        std::forward_as_tuple(location_ref(), intern(name), group,
+                              otf2::definition::location::location_type::cpu_thread));
+    return archive()(location.first->second);
 }
 
 otf2::writer::local& Trace::metric_writer(pid_t pid, pid_t tid)
 {
     auto name = (boost::format("metrics for thread %d") % tid).str();
-    auto location =
-        locations_.emplace(location_ref(), intern(name), location_groups_process_.at(pid),
-                           otf2::definition::location::location_type::metric);
-    return archive()(location);
+    // As the tid is unique in this context, create only one
+    // writer/location per tid
+    auto location = metric_locations_.emplace(
+        std::piecewise_construct, std::forward_as_tuple(tid),
+        std::forward_as_tuple(location_ref(), intern(name), location_groups_process_.at(pid),
+                              otf2::definition::location::location_type::metric));
+    return archive()(location.first->second);
 }
-
 otf2::writer::local& Trace::metric_writer(const std::string& name)
```

Comment:
Please give the code some space to breath :) (aka insert a newline between the two functions)